### PR TITLE
Fix KeyError on releasing resources of a Connection when closing the DB.

### LINKED ISSUE #208: intermittent KeyError on releasing resources of a Connection
When running tests (in plone.restapi) we sometimes get the following KeyError. It is related to timing issues - inserting a time.sleep(2) in the test setup causes the problem to occur more frequently.

It happens in plone.coredev using ZODB 5.4.0
It does not happen with ZODB 5.3.0

I think it is related to this commit: https://github.com/zopefoundation/ZODB/commit/1b9475d413705e510ff2f19ee6449688ffb04d46

In line 948 we see: `c.close(False)`

This is part of the traceback: 
`  File "/work/buildout_cache/eggs/ZODB-5.4.0-py2.7.egg/ZODB/Connection.py", line 948, in _release_resources
    c.close(False)
`
In 5.3.0 the corresponding line just sets the transaction manager to None:
`c.transaction_manager = None`

```
    20/20 (100.0%) test_total_item_count_is_correct (plone.restapi.tests.test_batching.TestBatchingSiteRoot) (1.888 s)Traceback (most recent call last):    
  File "./bin/test", line 445, in <module>
    '--test-path', '/work/playground/plone/plone52devel/src/ipdb',
  File "/work/buildout_cache/eggs/collective.xmltestreport-1.3.4-py2.7.egg/collective/xmltestreport/runner.py", line 60, in run
    failed = run_internal(defaults, args, script_parts=script_parts)
  File "/work/buildout_cache/eggs/collective.xmltestreport-1.3.4-py2.7.egg/collective/xmltestreport/runner.py", line 73, in run_internal
    runner.run()
  File "/work/buildout_cache/eggs/zope.testrunner-4.8.1-py2.7.egg/zope/testrunner/runner.py", line 189, in run
    self.run_tests()
  File "/work/buildout_cache/eggs/zope.testrunner-4.8.1-py2.7.egg/zope/testrunner/runner.py", line 291, in run_tests
    self.skipped, self.import_errors)
  File "/work/buildout_cache/eggs/zope.testrunner-4.8.1-py2.7.egg/zope/testrunner/runner.py", line 460, in run_layer
    import_errors)
  File "/work/buildout_cache/eggs/zope.testrunner-4.8.1-py2.7.egg/zope/testrunner/runner.py", line 383, in run_tests
    test(result)
  File "/usr/lib/python2.7/unittest/case.py", line 393, in __call__
    return self.run(*args, **kwds)
  File "/usr/lib/python2.7/unittest/case.py", line 370, in run
    result.stopTest(self)
  File "/work/buildout_cache/eggs/zope.testrunner-4.8.1-py2.7.egg/zope/testrunner/runner.py", line 897, in stopTest
    self.testTearDown()
  File "/work/buildout_cache/eggs/zope.testrunner-4.8.1-py2.7.egg/zope/testrunner/runner.py", line 821, in testTearDown
    layer.testTearDown()
  File "/work/playground/plone/plone52devel/src/plone.app.testing/plone/app/testing/layers.py", line 288, in testTearDown
    super(PloneTestLifecycle, self).testTearDown()
  File "/work/playground/plone/plone52devel/src/plone.testing/src/plone/testing/z2.py", line 945, in testTearDown
    self['zodbDB'].close()
  File "/work/buildout_cache/eggs/ZODB-5.4.0-py2.7.egg/ZODB/DB.py", line 651, in close
    @self._connectionMap
  File "/work/buildout_cache/eggs/ZODB-5.4.0-py2.7.egg/ZODB/DB.py", line 518, in _connectionMap
    self.pool.map(f)
  File "/work/buildout_cache/eggs/ZODB-5.4.0-py2.7.egg/ZODB/DB.py", line 213, in map
    self.all.map(f)
  File "/work/buildout_cache/eggs/transaction-2.2.1-py2.7.egg/transaction/weakset.py", line 62, in map
    f(elt)
  File "/work/buildout_cache/eggs/ZODB-5.4.0-py2.7.egg/ZODB/DB.py", line 659, in _
    conn._release_resources()
  File "/work/buildout_cache/eggs/ZODB-5.4.0-py2.7.egg/ZODB/Connection.py", line 948, in _release_resources
    c.close(False)
  File "/work/buildout_cache/eggs/ZODB-5.4.0-py2.7.egg/ZODB/Connection.py", line 304, in close
    self.transaction_manager.unregisterSynch(self)
  File "/work/buildout_cache/eggs/transaction-2.2.1-py2.7.egg/transaction/_manager.py", line 106, in unregisterSynch
    self._synchs.remove(synch)
  File "/work/buildout_cache/eggs/transaction-2.2.1-py2.7.egg/transaction/weakset.py", line 51, in remove
    del self.data[id(obj)]
  File "/usr/lib/python2.7/weakref.py", line 86, in __delitem__
    del self.data[key]
KeyError: 139676020831248


```

### PR DESCRIPTION
**Pull request by icemac**
Fixes #208 

Currently needs zopefoundation/transaction#68 to be merged and released.

### CODE REVIEW

File: src/ZODB/Connection.py [hunk 1]
```diff
@@ -884,7 +884,10 @@ def open(self, transaction_manager=None, delegate=True):
         """
 
         if transaction_manager is None:
-            transaction_manager = transaction.manager
+            # The .manager bit below unwraps the threaded
+            # manager so we can call unregisterSynch in close
+            # when close is called from another thread.
+            transaction_manager = transaction.manager.manager
```
Comment (mgedmin) [inline src/ZODB/Connection.py]: This sort of peeking into the implementation details of another package demands a versioned dependency requirement in `setup.py` IMHO.

**Comment by icemac** [comment]
@mgedmin Are you okay with my changes?

**Comment by mgedmin** [comment]
I do not fully understand what is going on here.

The title of this PR speaks about fixing a bug, but the changes themselves seem like this PR is adjusting for an implementation detail change in `transaction`.

Was the bug introduced by new `transaction`?  Was this an older bug that was fixed but new transaction broke the fix?  Was the older bug fixed without adding a regression test that would've caught the regression?  I'm happy to see that this PR adds a test (although I wish ZODB would use standard unittest conventions instead of `check_...` and manual test class enumeration).

I'm not sure why this API regression wouldn't be better fixed by making `transaction` forward the `unregisterSynch()` call to the underlying real transaction manager.  I believe @jamadden suggested forwarding all the methods using `getattr()`.  (I also believe @jimfulton made sure all public ITransactionManager methods were correctly forwarded, so why isn't unregisterSynch()?  Is it not public?  Is it part of some other interface?)

In short, I don't quite understand what is happening or what should be happening, and I don't currently have the time (or, to be honest, the interest) to study the problem and the solution, and as such I'll leave it to somebody else to approve the PR.

**Comment by mgedmin** [comment]
Wait, I think I get it now.

The code is not calling `manager.unregisterSynch()`.  It's passing the `manager` to `connection.unregisterSynch()`, and so it's important to pass the right `manager`.

**Comment by icemac** [comment]
@jimfulton @tseaver Can we have a new ZODB release including this fix?

**Comment by tseaver** [comment]
@icemac [5.5.1 release distributions on PyPY](https://pypi.org/project/ZODB/5.5.1/#files).

**Comment by icemac** [comment]
@tseaver Thank you very much! 🎉
